chainer/functions/tensor_network.py

```diff
@@ -0,0 +1,285 @@
+            # initializing weights in tensor network.
+            # This initialization is a modification of
+            # that of Linear function.
+            in_size = numpy.prod(self.in_sizes)
+            self.W = numpy.random.normal(
+                0, math.sqrt(1. / in_size),
+                (self.in_sizes[0], self.in_sizes[1], out_size)
+            ).astype(numpy.float32)
+
+        if not self.nobias:
+            if initial_bias is not None:
+                assert len(initial_bias) == 3
+                assert initial_bias[0].shape == (self.in_sizes[0], out_size)
+                assert initial_bias[1].shape == (self.in_sizes[1], out_size)
+                assert initial_bias[2].shape == (out_size,)
+                self.V1, self.V2, self.b = initial_bias
+            else:
+                self.V1 = numpy.zeros(
+                    (self.in_sizes[0], out_size), dtype=numpy.float32)
+                self.V2 = numpy.zeros(
+                    (self.in_sizes[1], out_size), dtype=numpy.float32)
+                self.b = numpy.zeros((out_size, ), dtype=numpy.float32)
+
+        self.gW = array.empty_like(self.W)
+        if not self.nobias:
+            self.gV1 = array.empty_like(self.V1)
+            self.gV2 = array.empty_like(self.V2)
+            self.gb = array.empty_like(self.b)
+
+    @property
+    def parameter_names(self):
+        if self.nobias:
+            return 'W',
+        assert self.V1 is not None
+        assert self.V2 is not None
+        assert self.b is not None
+        return 'W', 'V1', 'V2', 'b'
+
+    @property
+    def gradient_names(self):
+        if self.nobias:
+            return 'gW',
+        assert self.gV1 is not None
+        assert self.gV2 is not None
+        assert self.gb is not None
+        return 'gW', 'gV1', 'gV2', 'gb'
+
+    def check_type_forward(self, in_types):
+        type_check.expect(in_types.size() == 2)
+
+        e1_type, e2_type = in_types
+        type_check.expect(
+            e1_type.ndim >= 2,
+            e2_type.ndim >= 2,
+            e1_type.shape[0] == e2_type.shape[0]
+        )
+
+        in_sizes = type_check.Variable(self.in_sizes, 'in_sizes')
+        type_check_prod = type_check.Variable(numpy.prod, 'prod')
+        type_check.expect(
+            type_check_prod(e1_type.shape[1:]) == in_sizes[0],
+            type_check_prod(e2_type.shape[1:]) == in_sizes[1]
+        )
+
+    def zero_grads(self):
+        self.gW.fill(0)
+        if not self.nobias:
+            self.gV1.fill(0)
+            self.gV2.fill(0)
+            self.gb.fill(0)
+
+    def forward_cpu(self, x):
+        e1 = array.as_mat(x[0])
+        e2 = array.as_mat(x[1])
+        y = numpy.einsum('ij,ik,jkl->il', e1, e2, self.W)
+        if not self.nobias:
+            y += e1.dot(self.V1)
+            y += e2.dot(self.V2)
+            y += self.b
+        return y,
+
+    def forward_gpu(self, x):
+        i_len, j_len = x[0].shape
```

Comment:
I found `TensorNetwork` did work in the case `x[0].ndim >= 3`. So I fixed the code.